lib/MiLight/FUT089PacketFormatter.cpp

```diff
@@ -27,13 +27,73 @@ void FUT089PacketFormatter::updateColorRaw(uint8_t value) {
   command(FUT089_COLOR, FUT089_COLOR_OFFSET + value);
 }
 
+// change the temperature (kelvin).  Note that temperature and saturation share the same command 
+// number (7), and they change which they do based on the mode of the lamp (white vs. color mode).
+// To make this command work, we need to switch to white mode, make the change, and then flip
+// back to the original mode.
 void FUT089PacketFormatter::updateTemperature(uint8_t value) {
-  updateColorWhite();
+  // look up our current mode 
+  BulbId* ourBulb = new BulbId(this->deviceId, this->groupId, REMOTE_TYPE_FUT089);
+  GroupState ourState = this->stateStore->get(*ourBulb);
+  BulbMode originalBulbMode = ourState.getBulbMode();
+
+  // are we already in white?  If not, change to white
+  if (originalBulbMode != BulbMode::BULB_MODE_WHITE) {
+    updateColorWhite();
+  }
+
+  // now make the temperature change
   command(FUT089_KELVIN, 100 - value);
+
+  // revert back to the prior mode
+  switch (originalBulbMode) {
```

Comment:
I was actually just looking at the FUT089 formatter.  Seems like this code is here and in `updateSaturation`, with the caveat that the no-op case is different.  But seems like that'd be easy to abstract.

But you bring up a good point -- it actually seems like the base `PacketFormatter` should have an `switchBulbMode` method.  There are some leaky abstractions because not all device types support color, etc., but `PacketFormatter` already has `updateHue`.  What differences did you notice between formatters that made it seem like the abstraction would be difficult? 

I think you're right that it's not worth spending too much effort.

Reply:
I can look into abstracting those temperature/saturation state changes if you want.  If so, it would want to be added to the base class V2PacketFormatter so it can be shared across V2 devices (I guess it could go all the way back to PacketFormatter, but I'm not sure if this state issue exists in non-V2 devices?).  But I'd rather not try to refactor RgbCct and FUT089 into some common base class at this time.

If we want to look at this - do you know if V1 devices have any of these state issues?  I don't think so, but thought you might know better than I!